[clangd] Track IWYU pragmas for non-preamble includes

kadircet opened the pull request:
This makes PragmaIncldues copyable, and copies it from preamble when building a
new AST.

Fixes https://github.com/clangd/clangd/issues/1843
Fixes https://github.com/clangd/clangd/issues/1571

kadircet committed:
[clangd] Track IWYU pragmas for non-preamble includes

sam-mccall commented on clang-tools-extra/include-cleaner/include/clang-include-cleaner/Record.h:
```diff
@@ -113,7 +113,7 @@ class PragmaIncludes {
 
   /// Owns the strings.
-  llvm::BumpPtrAllocator Arena;
+  std::shared_ptr<llvm::BumpPtrAllocator> Arena;
```
This doesn't look threadsafe. Two PIs associated with parallel ParsedAST builds will write into the same arena.

You could do some trick with shared_ptr<void> to keep all needed arenas alive, and a raw pointer to the current, I guess?

kadircet replied:
you're right in theory, but in practice we actually never build ASTs concurrently from a preamble in clangd (we've a single ASTWorker peer per preamble). Moreover mutating the same arena bit is actually never the case, but quite subtle. We build the arena specifically in PPCallbacks and just move it to PI at the end. Hence in practice we have 2 separate PIs, one inside preamble, whose arena is never mutated. Another in ParsedAST, which can refer to strings both inside its own Arena, but also to the arena "kept" alive by the preamble. so I think this was safe for clangd, but too subtle to keep working in the long run I guess.

storing a vector of const arenas in PI instead, and changing the `record` to just push each arena to the end, rather than overwriting the existing Arena (while extending all other structures).

kadircet committed:
Make sure PI keeps old strings alive

Indicate that once an allocator is put into PI it is immutable, and also
make sure all the previous allocators out-lives the PI they were
recorded on.

hokein reviewed (approved):
thanks, looks good.
